pkg/autodiscovery/providers/config_reader.go

```diff
@@ -0,0 +1,418 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package providers
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/configresolver"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	cache "github.com/patrickmn/go-cache"
+	"gopkg.in/yaml.v2"
+)
+
+type configFormat struct {
+	ADIdentifiers           []string                           `yaml:"ad_identifiers"`
+	AdvancedADIdentifiers   []integration.AdvancedADIdentifier `yaml:"advanced_ad_identifiers"`
+	ClusterCheck            bool                               `yaml:"cluster_check"`
+	InitConfig              interface{}                        `yaml:"init_config"`
+	MetricConfig            interface{}                        `yaml:"jmx_metrics"`
+	LogsConfig              interface{}                        `yaml:"logs"`
+	Instances               []integration.RawMap
+	DockerImages            []string `yaml:"docker_images"`             // Only imported for deprecation warning
+	IgnoreAutodiscoveryTags bool     `yaml:"ignore_autodiscovery_tags"` // Use to ignore tags coming from autodiscovery
+}
+
+type configPkg struct {
+	confs    []integration.Config
+	defaults []integration.Config
+	others   []integration.Config
+}
+
+type configEntry struct {
+	conf       integration.Config
+	name       string
+	isDefault  bool
+	isMetric   bool
+	isLogsOnly bool
+	err        error
+}
+
+var reader *configFilesReader
+
+type configFilesReader struct {
+	paths []string
+	cache *cache.Cache
+	sync.Mutex
+}
+
+var doOnce sync.Once
+
+// InitConfigFilesReader initializes the config files reader.
+// It reads all configs and caches them in memory for 5 minutes.
+// InitConfigFilesReader should be called at agent startup.
+func InitConfigFilesReader(paths []string) {
+	doOnce.Do(func() {
```

Comment:
Since this is in a function that's only called once, it shouldn't need a `sync.Once`.

Reply:
It's correct that currently it's only called but I put this `sync.Once` as a safeguard for the future, it's not harmful to keep it and it may protect us in the future :)